zerver/views/auth.py

```diff
@@ -96,28 +97,26 @@ def get_safe_redirect_to(url: str, redirect_host: str) -> str:
     else:
         return redirect_host
 
-def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpResponse:
-    try:
-        email_change_object = get_object_from_key(confirmation_key, Confirmation.EMAIL_CHANGE)
-    except ConfirmationKeyException as exception:
-        return render_confirmation_key_error(request, exception)
-
+def confirm_email_change(request: HttpRequest, email_change_object: EmailChangeStatus,
+                         acting_user: UserProfile) -> HttpResponse:
     new_email = email_change_object.new_email
     old_email = email_change_object.old_email
-    user_profile = email_change_object.user_profile
 
-    if user_profile.realm.email_changes_disabled and not user_profile.is_realm_admin:
+    if acting_user != email_change_object.user_profile:
+        raise JsonableError(_("Incorrect login. Login with your old email to change your email address"))
+
```

Comment:
let's just set `user_profile = email_change_object.user_profile` here after the validation annd keep using that vairable below.